blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java

```diff
@@ -27,23 +43,51 @@ public void onCompleted(WorkflowRun workflowRun, @Nonnull TaskListener listener)
         if (analytics == null) {
             return;
         }
+        FlowExecution execution = workflowRun.getExecution();
+        if (execution == null) {
+            return;
+        }
         // Tally up all the steps used in this run
-        Tally tally = new Tally();
-        NodeGraphBuilder builder = NodeGraphBuilder.NodeGraphBuilderFactory.getInstance(workflowRun);
-        builder.getPipelineNodeSteps(new Link("steps/")).forEach(step -> {
-            tally.count(step.getStepType());
-        });
+        PipelineStepVisitor visitor = new PipelineStepVisitor();
+        ForkScanner.visitSimpleChunks(execution.getCurrentHeads(), visitor, new StageChunkFinder());
+
         boolean isDeclarative = workflowRun.getParent().getAction(DeclarativeJobAction.class) != null;
         Result result = workflowRun.getResult();
         String resultAsString = result != null ? result.toString() : "UNKNOWN";
         // Send event for each step used in this run
-        tally.get().forEach((key, value) -> {
+        visitor.tally.get().forEach((key, value) -> {
             ImmutableMap.Builder<String, Object> props = ImmutableMap.builder();
             props.put("type", key);
             props.put("timesUsed", value);
             props.put("isDeclarative", isDeclarative);
             props.put("runResult", resultAsString);
+            String pluginName = visitor.stepToPlugin.get(key);
+            props.put("plugin", pluginName != null ? pluginName : "unknown");
             analytics.track(new TrackRequest("pipeline_step_used", props.build()));
         });
     }
+
+    private static final class PipelineStepVisitor extends StandardChunkVisitor {
```

Comment:
If you're just collecting a Tally, use https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowNodeVisitor.java and 
https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java#L373

Bismuth is meant for this kind of work, and meant to make it simple -- you need not just collect on AtomNodes, we can get info on which block-scoped steps were run as well -- caveat: you might get some duplication due to nesting of blocks.

I'd actually imagined doing counts of steps by type but nobody asked for it, so I didn't code it up.

Reply:
On second thought, better approach here is probably to use a GraphListener so you're not doing a complete extra scan through the FlowGraph.